### DISCUSSION TITLE
Researchspace with inference?

### DISCUSSION THREAD

**Discussion by julianklotz** [researchspace/researchspace#268] [Q&A] [answered]
Dear ResearchSpace community,

I’m struggling to set up inferencing for ResearchSpace and Blazegraph.

Here’s my situation:

My organization has its own ontology (prefix mo:) that is built upon CIDOC and FRBRoo. To simplify using both ontologies, mo: uses subclasses & subproperties to create a domain-specific version of both.

We use the mo ontology to build a graph from our own data, something like this:
```
# Imports skipped for brevity

# Ontology-level
mo a owl:Ontology;
mo:Performance rdfs:subClassOf frbroo:F31_Performance .

# Data-level
example:a_performance a mo:Performance ;
    rdfs:label "An example performance." .
```
ResearchSpace doesn’t have inferencing enabled by default, so it will show nothing more than sample data. To benefit from Researchspace, I have to use inferencing at some point.

I have tried three options:

1. Do inferencing at ontology build time, i.e. run the mo ontology through a reasoner before adding ontology + data to Researchspace using the data upload. This approach works, but doesn't allow automatic inferencing when adding data on a daily basis.
2. Connect our own GraphDB repository that has inferencing enabled, using the [Repository manager](http://localhost:10214/resource/Admin:Repositories). Technically speaking, it works. I can query the SPARQL endpoint from ResearchSpace’s SPARQL console by selecting the right repository. Inferenced triples show up just as expected. Though, it seems all queries done through the UI (e.g. Thinking Frames) seem to use the `default` repository. There is no setting and no documentation I could find that lets me change this behavior. The documentation on the [Repository Manager](https://documentation.researchspace.org/resource/Help:WorkingWithData) doesn't talk about how the *use* the configured repositories at all.
3. Enable BlazeGraph inferencing. Using the standalone MacOS version, I changed `webapps-exploded/blazegraph/webapp/WEB-INF/GraphStore.properties` before first startup. It looks like this now:
```
[…]
##
## Setup for triples mode + full text search but ~~without~~ with inference.
##
com.bigdata.rdf.sail.truthMaintenance=true
com.bigdata.rdf.store.AbstractTripleStore.quads=true
com.bigdata.rdf.store.AbstractTripleStore.statementIdentifiers=true
com.bigdata.rdf.store.AbstractTripleStore.textIndex=true
# com.bigdata.rdf.store.AbstractTripleStore.axiomsClass=com.bigdata.rdf.axioms.NoAxioms
com.bigdata.rdf.store.AbstractTripleStore.rdrHistoryClass=com.bigdata.rdf.sail.RDRHistory
```
Still, only explicit statements show up.

Except option (1), none of this really worked. Can anyone point me in the right direction? Any help is appreciated.

Thanks,
Julian

**Comment by julianklotz** [comment]
Bump

**Comment by lklic** [comment, marked as answer]
Hi,
For option 2, you can set GraphDB as the default repository. Just change the values of the default repository to point to your GraphDB instance. At that point you can turn off your blazegraph instance (if it is still on). By default, the `assets` repository is where knowledge patterns/field definitions are stored, and this is normally proxied to the default repository, but it can be changed to use the RDF4J (i.e. the `native` template you can see in the repository manager), or any other graph store.

for option 3, I haven't tested blazegraph inferencing so I am not sure how well it works. But I think you will have to delete the default `kb` namespace in the blazegraph web UI and add it again with quads turned off, and inferencing turned on. By turning off quads mode, I think you will lose a lot of functionality however, and it may not work with RS. For example, LDP containers use named graphs to identify the container.

Would be curious to know how what solution you end up with!

**Reply by julianklotz** [reply]
Thanks! I’ll check this out.

**Reply by sopheck** [reply]
Hey @julianklotz , curious if you've made any progress regarding blazegraph inferencing? Thx

**Reply by julianklotz** [reply]
Hi @sopheck,

to be honest, it neither worked with Blazegraph nor GraphDB.

Judging from the repo on Github, Blazegraph has been abandoned. Its team has moved on to Amazon Neptune and I don’t expect to see much activity here in the future. Not long after asking this question, we gave up on experimenting with ResearchSpace, because (from what I can see from the outside) it isn’t actively maintained.

Sorry for this not-so-helpful reply!
